src/Workspaces/Core/Portable/Workspace/Solution/StateChecksums.cs

```diff
@@ -2,273 +2,374 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+using System;
 using System.Collections.Generic;
-using System.Collections.Immutable;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
 using System.Runtime.CompilerServices;
 using System.Threading;
 using System.Threading.Tasks;
 using Roslyn.Utilities;
 
-namespace Microsoft.CodeAnalysis.Serialization
+namespace Microsoft.CodeAnalysis.Serialization;
+
+internal sealed class SolutionStateChecksums(
+    Checksum attributesChecksum,
+    ChecksumCollection projectChecksums,
+    ChecksumCollection analyzerReferenceChecksums,
+    Checksum frozenSourceGeneratedDocumentIdentity,
+    Checksum frozenSourceGeneratedDocumentText) : IChecksummedObject
 {
-    internal sealed class SolutionStateChecksums(ImmutableArray<object> children) : ChecksumWithChildren(children)
+    public Checksum Checksum { get; } = Checksum.Create(stackalloc[]
+    {
+        attributesChecksum.Hash,
+        projectChecksums.Checksum.Hash,
+        analyzerReferenceChecksums.Checksum.Hash,
+        frozenSourceGeneratedDocumentIdentity.Hash,
+        frozenSourceGeneratedDocumentText.Hash,
+    });
+
+    public Checksum Attributes => attributesChecksum;
+    public ChecksumCollection Projects => projectChecksums;
+    public ChecksumCollection AnalyzerReferences => analyzerReferenceChecksums;
+    public Checksum FrozenSourceGeneratedDocumentIdentity => frozenSourceGeneratedDocumentIdentity;
+    public Checksum FrozenSourceGeneratedDocumentText => frozenSourceGeneratedDocumentText;
+
+    public void AddAllTo(HashSet<Checksum> checksums)
+    {
+        checksums.Add(this.Checksum);
+        checksums.Add(this.Attributes);
+        this.Projects.AddAllTo(checksums);
+        this.AnalyzerReferences.AddAllTo(checksums);
+        checksums.Add(this.FrozenSourceGeneratedDocumentIdentity);
+        checksums.Add(this.FrozenSourceGeneratedDocumentText);
+    }
+
+    public void Serialize(ObjectWriter writer)
+    {
+        // Writing this is optional, but helps ensure checksums are being computed properly on both the host and oop side.
+        this.Checksum.WriteTo(writer);
```

Comment:
I think a bit more verbostity and congruency between WriteTo/ReadFrom would have made it easier for me to read this, eg:

```
    public void Serialize(ObjectWriter writer)
    {
        // Writing this is optional, but helps ensure checksums are being computed properly on both the host and oop side.
        Checksum.WriteTo(this.Checksum, writer);

        Checksum.WriteTo(this.Attributes, writer);
        ChecksumCollection.WriteTo(this.Projects, writer);
        ChecksumCollection.WriteTo(this.AnalyzerReferences, writer);
        Checksum.WriteTo(this.FrozenSourceGeneratedDocumentIdentity, writer);
        Checksum.WriteTo(this.FrozenSourceGeneratedDocumentText, writer);
    }

    public static SolutionStateChecksums Deserialize(ObjectReader reader)
    {
        var checksum = Checksum.ReadFrom(reader);

        var attributesChecksum = Checksum.ReadFrom(reader);
        var projectsChecksum = ChecksumCollection.ReadFrom(reader);
        var analyzerReferencesChecksum = ChecksumCollection.ReadFrom(reader);
        var frozenSourceGeneratedDocumentIdentityChecksum = Checksum.ReadFrom(reader);
        var frozenSourceGeneratedDocumentTextChecksum = Checksum.ReadFrom(reader);

        var result = new SolutionStateChecksums(
            attributesChecksum,
            projectsChecksum,
            analyzerReferencesChecksum,
            frozenSourceGeneratedDocumentIdentityChecksum,
            frozenSourceGeneratedDocumentTextChecksum);

        Contract.ThrowIfFalse(result.Checksum == checksum);
        return result;
    }
```

Reply:
I personally don't think that helps. I can use named params here though.

Reply:
updated to use named parameters.